File: project_euler/problem_044/solution.py
import math


def pent(n):
    """Compute the n-th pentagonal number P_n = n(3n-1)/2."""
    return n * (3 * n - 1) // 2


def is_pentagonal(v):
    """Test if v is pentagonal: v = P_n iff (1 + sqrt(1+24v)) / 6 is a
    positive integer, equivalently 1+24v is a perfect square s^2 with
    s = 5 mod 6.
    """
    if v <= 0:
        return False
    disc = 1 + 24 * v
    s = math.isqrt(disc)
    if s * s != disc:
        return False
    return (1 + s) % 6 == 0


def solve():
    """Find the pair (P_j, P_k) with j < k such that both P_k - P_j and
    P_k + P_j are pentagonal, and D = P_k - P_j is minimised.

    Search strategy: iterate k, for each k try j = k-1 downto 1.
    Terminate outer loop when the minimum gap 3k-2 exceeds current best.
    """
    best = float('inf')
    for k in range(2, 10000):
        pk = pent(k)
        if 3 * k - 2 >= best:
            break
        for j in range(k - 1, 0, -1):
            pj = pent(j)
            diff = pk - pj
            if diff >= best:
                break
            if is_pentagonal(diff) and is_pentagonal(pk + pj):
                best = diff
    print(best)


solve()

File: project_euler/problem_044/solution.cpp
#include <bits/stdc++.h>
using namespace std;

long long pent(long long n) {
    return n * (3 * n - 1) / 2;
}

bool is_pentagonal(long long v) {
    // v is pentagonal iff 1+24v = s^2 with s = 5 mod 6,
    // equivalently (1 + s) % 6 == 0.
    if (v <= 0) return false;
    long long disc = 1 + 24 * v;
    long long s = (long long)sqrt((double)disc);
    while (s * s < disc) s++;
    while (s * s > disc) s--;
    if (s * s != disc) return false;
    return (1 + s) % 6 == 0;
}

int main() {
    // Search for minimal D = P_k - P_j with both D and P_k + P_j pentagonal.
    // Terminate when minimum gap 3k-2 >= best.
    long long best = LLONG_MAX;
    for (long long k = 2; ; k++) {
        if (3 * k - 2 >= best) break;
        long long pk = pent(k);
        for (long long j = k - 1; j >= 1; j--) {
            long long pj = pent(j);
            long long diff = pk - pj;
            if (diff >= best) break;
            if (is_pentagonal(diff) && is_pentagonal(pk + pj)) {
                best = diff;
            }
        }
    }
    cout << best << endl;
    return 0;
}
